src/units.jl

```diff
@@ -0,0 +1,261 @@
+    vel_dim, vel_units = validate_velocities(velocities)
+    force_dim = dimension(force_units)
+    energy_dim = dimension(energy_units)
+    mass_dim, mass_units = validate_masses(masses)
+    validate_energy_units(energy_units)
+
+    forceIsMolar = (force_dim == u"𝐋 * 𝐌 * 𝐍^-1 * 𝐓^-2")
+    energyIsMolar = (energy_dim == u"𝐋^2 * 𝐌 * 𝐍^-1 * 𝐓^-2")
+    massIsMolar = (mass_dim == u"𝐌* 𝐍^-1")
+    
+    if !allequal([energyIsMolar, massIsMolar, forceIsMolar])
+        throw(ArgumentError("""System was constructed with inconsistent energy, force & mass units.\
+         All must be molar, non-molar or unitless. For example, kcal & kg are allowed but kcal/mol\
+         and kg is not allowed. Units were: $([energy_units, mass_units, force_units])"""))
+    end
+
+    no_dim_arr = [dim == NoDims for dim in [length_dim, vel_dim, energy_dim, force_dim, mass_dim]]
+
+    # If something has NoDims, all other data must have NoDims
+    if any(no_dim_arr) && !all(no_dim_arr)
+        throw(ArgumentError("""Either coords, velocities, masses or energy_units has NoDims/NoUnits but\
+         the others do have units. Molly does not permit mixing dimensionless and dimensioned data."""))
+    end
+
+    #Check derived units
+    if force_units != (energy_units / length_units)
+        throw(ArgumentError("Force unit was specified as $(force_units), but that unit could not be re-derived
+            from the length units in coords and the energy_units passed to `System`"))
+    end
+
+    #TODO: How do I check velocity units are length_units / time and not something stupid???
+    #TODO: time unit can be whatever, I just want to avoid velo being sqrt(J/kg)
+    # if vel_units != (length_units / ??)
+
+    # end
+
+    return NamedTuple{(:length, :velocity, :mass, :energy, :force)}((length_units,
+        vel_units, mass_units, energy_units, force_units))
+
+end
+
+#TODO: THIS HAS ISSUES BECAUISE SOME OF THE INTERS DONT DEFINE THESE? Best we can do for now?
+function check_interaction_units(p_inters, s_inters, g_inters, sys_units::NamedTuple)
+
+    for inter in [p_inters; s_inters; g_inters]
+        if hasproperty(inter, :energy_units)
+            if inter.energy_units != sys_units[:energy]
+                throw(ArgumentError("Energy units passed to system do not match those passed to interactions"))
+            end
+        end
+
+        if hasproperty(inter, :force_units)
+            if inter.force_units != sys_units[:force]
+                throw(ArgumentError("Force units passed to system do not match those passed to interactions"))
+            end
+        end
+    end
+
+end
+
+function check_other_units(atoms, boundary, constraints, sys_units::NamedTuple)
+    box_units = unit(boundary)
+
+    if !all(sys_units[:length] .== box_units)
+        throw(ArgumentError("Simulation box constructed with $(box_units) but length unit on coords was $(sys_units[:length])"))
+    end
+
+    σ_units = unit.(getproperty.(atoms, :σ))
```

Comment:
These are not required properties for atoms, so this could be inside a `hasproperty` check.

Reply:
hasproperty will always be true because there are default values though. I do agree if they are unused they should not be checked, but the real fix for that is just not having them in the Atom constructor. I'll just check if they match the defaults for now to determine whether or not I check the units.

Reply:
Simulations don't have to be run with the `Atom` type, they can be run with any type. Other types may not have those properties.